src/array/cpu/cpu_id_hash_map.cc

```diff
@@ -0,0 +1,194 @@
+/**
+ *  Copyright (c) 2019 by Contributors
+ * @file array/cpu/cpu_id_hash_map.cc
+ * @brief Class about CPU id hash map
+ */
+
+#include "cpu_id_hash_map.h"
+
+#include <dgl/runtime/parallel_for.h>
+#include <dgl/runtime/device_api.h>
+
+#include <cmath>
+
+using namespace dgl::runtime;
+
+namespace dgl {
+namespace aten {
+
+template <typename IdType>
+IdType CpuIdHashMap<IdType>::CompareAndSwap(IdType* ptr,
+    IdType old_val, IdType new_val) {
+  #ifdef _MSC_VER
+  if (sizeof(IdType) == 4) {
+    return _InterlockedCompareExchange(reinterpret_cast<LONG*>(ptr),
+      new_val, old_val);
+  } else if (sizeof(IdType) == 8) {
+    return _InterlockedCompareExchange64(reinterpret_cast<LONGLONG*>(ptr),
+      new_val, old_val);
+  } else {
+    LOG(FATAL) << "ID can only be int32 or int64";
+  }
+  #elif __GNUC__   // _MSC_VER
+    return __sync_val_compare_and_swap(ptr, old_val, new_val);
```

Comment:
why turn to these built-in CAS instead of previous used 1std:;atomic<>::compare_exchange_weak1 ?

Reply:
Atomic is redundant and time-cousuming to init.

Reply:
> Atomic is redundant and time-consuming to init.